File: packages/ai-config/src/runtime/resolveEffectiveModel.ts
import { materializeRuntimeConfig } from '../io/configLoader.js';
import { resolveAuthSecret, type SecretResolverContext } from '../secrets/resolver.js';
import type { ProviderDiscoveryPlugin } from '../discovery/pluginInterface.js';
import { runProviderDiscovery } from '../discovery/discoveryEngine.js';
import { mergeCatalogs } from '../catalog/merge.js';
import { normalizeCatalog } from '../catalog/normalize.js';
import { selectModel, type SelectionInput } from '../selection/selector.js';
import type { Catalog } from '../domain/catalog.js';

export interface ResolveEffectiveModelContext extends SelectionInput, SecretResolverContext {
  plugins?: ProviderDiscoveryPlugin[];
}

export async function resolveEffectiveModel(rawConfig: unknown, context: ResolveEffectiveModelContext = {}) {
  const runtimeConfig = materializeRuntimeConfig(rawConfig, { env: context.env });

  const explicitCatalog: Catalog = {
    providers: runtimeConfig.providers.map((provider) => ({
      ...provider,
      auth: resolveAuthSecret(provider.auth, provider.id, {
        env: context.env,
        profiles: runtimeConfig.profiles,
        profileName: context.profileName,
      }),
    })),
    models: runtimeConfig.models,
    mode: runtimeConfig.mode,
    defaults: {
      provider: runtimeConfig.defaults.provider,
      model: runtimeConfig.defaults.model,
      byContext: runtimeConfig.defaults.byContext,
    },
  };

  const discovered = await runProviderDiscovery(context.plugins ?? [], {
    profileName: context.profileName,
    env: context.env,
  });

  const merged = mergeCatalogs(explicitCatalog, discovered);
  const normalized = normalizeCatalog(merged);

  return {
    catalog: normalized,
    selection: selectModel(normalized, {
      requested: context.requested,
      contextId: context.contextId,
      fallback: context.fallback,
      allowlist: context.allowlist ?? runtimeConfig.allowlist,
    }),
  };
}

File: packages/ai_config/runtime/resolve_effective_model.py
from typing import Dict, List, Optional

from ..catalog.merge import merge_catalogs
from ..catalog.normalize import normalize_catalog
from ..discovery.discovery_engine import run_provider_discovery
from ..io.config_loader import materialize_runtime_config
from ..secrets.resolver import resolve_auth_secret
from ..selection.selector import select_model


def resolve_effective_model(raw_config: Dict, context: Optional[Dict] = None) -> Dict:
    context = context or {}
    runtime_config = materialize_runtime_config(raw_config, env=context.get("env"))

    providers = []
    for provider in runtime_config.providers:
        resolved_auth = resolve_auth_secret(
            provider.auth,
            provider.id,
            env=context.get("env"),
            profiles=runtime_config.profiles,
            profile_name=context.get("profile_name"),
        )
        provider_dict = provider.model_dump() if hasattr(provider, "model_dump") else provider.dict()
        provider_dict["auth"] = resolved_auth.model_dump() if hasattr(resolved_auth, "model_dump") else resolved_auth.dict()
        providers.append(provider_dict)

    explicit_catalog = {
        "providers": providers,
        "models": [m.model_dump() if hasattr(m, "model_dump") else m.dict() for m in runtime_config.models],
        "mode": runtime_config.mode,
        "defaults": {
            "provider": runtime_config.defaults.provider,
            "model": runtime_config.defaults.model,
            "by_context": {
                key: value.model_dump() if hasattr(value, "model_dump") else value.dict()
                for key, value in runtime_config.defaults.by_context.items()
            },
        },
    }

    discovered = run_provider_discovery(context.get("plugins", []), {
        "profile_name": context.get("profile_name"),
        "env": context.get("env"),
    })

    merged = merge_catalogs(explicit_catalog, discovered)
    normalized = normalize_catalog(merged)

    selection = select_model(
        normalized,
        requested=context.get("requested"),
        context_id=context.get("context_id"),
        fallback=context.get("fallback"),
        allowlist=context.get("allowlist") or runtime_config.allowlist,
    )

    return {"catalog": normalized, "selection": selection}
